src/main/java/soot/baf/toolkits/base/OpStackCalculator.java

```diff
@@ -0,0 +1,863 @@
+import soot.LongType;
+import soot.PatchingChain;
+import soot.RefLikeType;
+import soot.RefType;
+import soot.SootMethod;
+import soot.StmtAddressType;
+import soot.Trap;
+import soot.Type;
+import soot.Unit;
+import soot.VoidType;
+import soot.baf.AddInst;
+import soot.baf.AndInst;
+import soot.baf.ArrayLengthInst;
+import soot.baf.ArrayReadInst;
+import soot.baf.ArrayWriteInst;
+import soot.baf.BafBody;
+import soot.baf.CmpInst;
+import soot.baf.CmpgInst;
+import soot.baf.CmplInst;
+import soot.baf.DivInst;
+import soot.baf.Dup1Inst;
+import soot.baf.Dup1_x1Inst;
+import soot.baf.Dup1_x2Inst;
+import soot.baf.Dup2Inst;
+import soot.baf.Dup2_x1Inst;
+import soot.baf.Dup2_x2Inst;
+import soot.baf.DynamicInvokeInst;
+import soot.baf.EnterMonitorInst;
+import soot.baf.ExitMonitorInst;
+import soot.baf.FieldGetInst;
+import soot.baf.FieldPutInst;
+import soot.baf.GotoInst;
+import soot.baf.IdentityInst;
+import soot.baf.IfCmpEqInst;
+import soot.baf.IfCmpGeInst;
+import soot.baf.IfCmpGtInst;
+import soot.baf.IfCmpLeInst;
+import soot.baf.IfCmpLtInst;
+import soot.baf.IfCmpNeInst;
+import soot.baf.IfEqInst;
+import soot.baf.IfGeInst;
+import soot.baf.IfGtInst;
+import soot.baf.IfLeInst;
+import soot.baf.IfLtInst;
+import soot.baf.IfNeInst;
+import soot.baf.IfNonNullInst;
+import soot.baf.IfNullInst;
+import soot.baf.IncInst;
+import soot.baf.Inst;
+import soot.baf.InstSwitch;
+import soot.baf.InstanceCastInst;
+import soot.baf.InstanceOfInst;
+import soot.baf.InterfaceInvokeInst;
+import soot.baf.JSRInst;
+import soot.baf.LoadInst;
+import soot.baf.LookupSwitchInst;
+import soot.baf.MethodArgInst;
+import soot.baf.MulInst;
+import soot.baf.NegInst;
+import soot.baf.NewArrayInst;
+import soot.baf.NewInst;
+import soot.baf.NewMultiArrayInst;
+import soot.baf.NopInst;
+import soot.baf.OpTypeArgInst;
+import soot.baf.OrInst;
+import soot.baf.PopInst;
+import soot.baf.PrimitiveCastInst;
+import soot.baf.PushInst;
+import soot.baf.RemInst;
+import soot.baf.ReturnInst;
+import soot.baf.ReturnVoidInst;
+import soot.baf.ShlInst;
+import soot.baf.ShrInst;
+import soot.baf.SpecialInvokeInst;
+import soot.baf.StaticGetInst;
+import soot.baf.StaticInvokeInst;
+import soot.baf.StaticPutInst;
+import soot.baf.StoreInst;
+import soot.baf.SubInst;
+import soot.baf.SwapInst;
+import soot.baf.TableSwitchInst;
+import soot.baf.TargetArgInst;
+import soot.baf.ThrowInst;
+import soot.baf.UshrInst;
+import soot.baf.VirtualInvokeInst;
+import soot.baf.XorInst;
+import soot.baf.internal.AbstractOpTypeInst;
+import soot.toolkits.graph.BriefUnitGraph;
+import soot.util.Chain;
+
+/**
+ * @author Michael Batchelder
+ * @author Timothy Hoffman
+ */
+public class OpStackCalculator {
+  private static final Logger logger = LoggerFactory.getLogger(OpStackCalculator.class);
+
+  private static class StackEffectSwitch implements InstSwitch {
+
+    public boolean shouldThrow = true;
+    public Type[] remove_types = null;
+    public Type[] add_types = null;
+
```

Comment:
What are the purpose of using these two Type arrays.

Reply:
They are basically gen and kill sets for the types on the JVM stack that is being emulated.

Reply:
add comments next to them or to the javadoc the class